In [9]:
import math
import geocoder
import folium
from geopy.distance import distance
from math import radians, sin, cos, sqrt, tan, atan, degrees

# Calculating the FOV of the Main Camera on Samsung Note 20 Ultra and Boundary Angles


## 1. Technical Data of the Main Camera
- **Focal length:** 26 mm (equivalent **35mm film format**, also known as **full-frame format** in digital photography.).
- To calculate the Field of View (FOV) for the Samsung Note 20 Ultra using the full-frame format equivalent, we will treat its reported focal lengths (**26mm for main wide**, 13mm for ultra-wide, and 120mm for telephoto) as "35mm equivalent focal lengths." This means we will use the standard **full-frame dimensions** of **36mm × 24mm** as the sensor size.

## 2. FOV Calculation Formula
$$
\text{FOV} = 2 \cdot \arctan\left(\frac{\text{sensor size}}{2 \cdot \text{focal length}}\right)
$$

### Horizontal FOV (hFOV):
$$
\text{hFOV} = 2 \cdot \arctan\left(\frac{\text{Sensor width}}{2 \cdot \text{focal length}}\right)
$$

### Vertical FOV (vFOV):
$$
\text{vFOV} = 2 \cdot \arctan\left(\frac{\text{sensor height}}{2 \cdot \text{focal length}}\right)
$$

### Diagonal FOV (dFOV):
$$
\text{dFOV} = 2 \cdot \arctan\left(\frac{\text{sensor diagonal}}{2 \cdot \text{focal length}}\right)
$$

In [29]:
def calculate_fov(sensor_size, focal_length):
    """
    Calculate the Field of View (FOV).

    :param sensor_size: Size of the sensor (in mm).
    :param focal_length: Focal length of the lens (in mm).
    :return: FOV in degrees.
    """
    fov = 2 * degrees(atan(sensor_size / (2 * focal_length)))
    return fov

In [58]:
# Sensor dimensions and focal length in mm
sensor_diagonal = sqrt(36**2 + 24**2)
sensor_ratio = (4, 3)   # Aspect ratio of the sensor
sensor_height = sensor_diagonal * (sensor_ratio[0] / sqrt(sensor_ratio[0]**2 + sensor_ratio[1]**2))
sensor_width = sensor_diagonal * (sensor_ratio[1] / sqrt(sensor_ratio[0]**2 + sensor_ratio[1]**2))
focal_length = 26

# Calculate FOVs
hfov = calculate_fov(sensor_width, focal_length)   # Horizontal FOV
vfov = calculate_fov(sensor_height, focal_length)  # Vertical FOV
dfov = calculate_fov(sensor_diagonal, focal_length)  # Diagonal FOV

# Print results
print("Horizontal FOV:", round(hfov, 2), "degrees")
print("Vertical FOV:", round(vfov, 2), "degrees")
print("Diagonal FOV:", round(dfov, 2), "degrees")

Horizontal FOV: 53.06 degrees
Vertical FOV: 67.3 degrees
Diagonal FOV: 79.52 degrees


If the camera uses a **1:1 aspect ratio**, it will indeed crop the image to create a square format. In this case:

- The **horizontal FOV (hFOV)** and **vertical FOV (vFOV)** will become equal because the sensor width and height are constrained to the same value (cropping the larger dimension to match the smaller one).

- Using the previously calculated **53.06° vFOV** for the main wide camera, we find that the **hFOV** and **vFOV** will both equal **53.06°**.

- The diagonal field of view (dFOV) can be calculated using the relationship between the horizontal, vertical, and diagonal dimensions of the cropped sensor:

$$
\text{dFOV} = 2 \times \arctan\left(\frac{\sqrt{(\text{Sensor Width})^2 + (\text{Sensor Height})^2}}{2 \times \text{Focal Length}}\right)
$$

$$
\text{dFOV} = 2 \times \arctan\left(\sqrt{2} \times \tan\left(\frac{\text{hFOV}}{2}\right)\right)
$$

In [63]:
hfov_1_1 = hfov
vfov_1_1 = hfov
dfov_1_1 = 2 * degrees(atan(sqrt(2) * tan(radians(hfov / 2))))

# Print results
print("Horizontal FOV in a 1:1 aspect ratio:", round(hfov_1_1, 2), "degrees")
print("Vertical FOV in a 1:1 aspect ratio:", round(vfov, 2), "degrees")
print("Diagonal FOV in a 1:1 aspect ratio:", round(dfov_1_1, 2), "degrees")

Horizontal FOV in a 1:1 aspect ratio: 53.06 degrees
Vertical FOV in a 1:1 aspect ratio: 67.3 degrees
Diagonal FOV in a 1:1 aspect ratio: 70.45 degrees


In [68]:
vfov_16_9 = vfov
sensor_width_16_9 = sensor_width * 9/12
hfov_16_9 = calculate_fov(sensor_width_16_9, focal_length)
#dfov_16_9 = 2 * degrees(atan(sqrt(2) * tan(radians(hfov / 2))))

# Print results
print("Horizontal FOV in a 16:9 aspect ratio:", round(hfov_16_9, 2), "degrees")
print("Vertical FOV in a 16:9 aspect ratio:", round(vfov_16_9, 2), "degrees")
#print("Diagonal FOV in a 16:9 aspect ratio:", round(dfov_16_9, 2), "degrees")

Horizontal FOV in a 16:9 aspect ratio: 41.05 degrees
Vertical FOV in a 16:9 aspect ratio: 67.3 degrees


### **Quy trình sử dụng Compass và chụp ảnh bầu trời**
1. **Xác định hướng (trước khi dựng điện thoại lên)**:
   - Đặt điện thoại **song song với mặt đất**.
   - Dùng Compass để ghi lại hướng trung tâm (θc), ví dụ: 45° (Đông Bắc).
   - Điều này giúp đảm bảo thông tin đầu vào chính xác cho việc tính toán góc biên trái và phải.

2. **Dựng điện thoại để chụp ảnh**:
   - Sau khi biết hướng trung tâm, bạn có thể dựng điện thoại theo góc phù hợp để chụp bầu trời.
   - Cách đặt:
     - **Thẳng đứng (90°)** nếu muốn tập trung hoàn toàn vào bầu trời.
     - **Nghiêng nhẹ** nếu muốn bao quát cả đường chân trời.

3. **Tính toán vùng quan sát**:
   - Dựa trên hướng trung tâm đo được và góc mở của camera (FOV), tính toán góc biên để xác định vùng bầu trời trong ảnh.

---

### **Lý do không dùng Compass khi điện thoại dựng đứng**
- Khi điện thoại dựng đứng, cảm biến từ trường đo hướng **theo phương dọc** không đáng tin cậy do nhiễu từ trường theo phương khác.
- Kết quả có thể sai lệch, ảnh hưởng đến độ chính xác khi tính toán góc biên.

---

**Tóm lại**: Đặt điện thoại **song song với mặt đất để đo hướng**, sau đó dựng lên để chụp ảnh. Điều này giúp đảm bảo cả dữ liệu đầu vào và bức ảnh đều chính xác.


## Calculating Left and Right Boundary Angles (Horizontal)
- **Central orientation (θc):** mesured by Compass.
- **Left boundary angle (θ_left):**
$$
\theta_{\text{left}} = \theta_c - \frac{\text{hFOV}}{2}
$$
- **Right boundary angle (θ_right):**
$$
\theta_{\text{right}} = \theta_c + \frac{\text{hFOV}}{2}
$$

In [31]:
def calculate_boundary_angles(center_angle, fov):
    """
    Calculate the left and right boundary angles based on the center angle and FOV.

    :param center_angle: Center orientation angle (in degrees).
    :param fov: Field of View (in degrees).
    :return: Left and right angles.
    """
    left_angle = center_angle - (fov / 2)
    right_angle = center_angle + (fov / 2)
    return left_angle, right_angle

In [32]:
# Center orientation angle (e.g., facing 45 degrees)
center_orientation = 310  # in degrees
print("Center orientation:", center_orientation, "degrees")

# Calculate left and right boundary angles
left_angle, right_angle = calculate_boundary_angles(center_orientation, hfov)

print("Left boundary angle:", round(left_angle, 2), "degrees")
print("Right boundary angle:", round(right_angle, 2), "degrees")

Center orientation: 310 degrees
Left boundary angle: 283.47 degrees
Right boundary angle: 336.53 degrees


#### **Thay đổi vĩ độ $( \Delta \text{lat} $):**
$$
\Delta \text{lat} = \frac{D}{R} \cdot \cos(\theta)
$$

#### **Thay đổi kinh độ ($ \Delta \text{lon} $):**
$$
\Delta \text{lon} = \frac{D}{R \cdot \cos(\text{latitude})} \cdot \sin(\theta)
$$

- $ D $: Khoảng cách quan sát tối đa (tính bằng mét).
- $ R $: Bán kính Trái Đất (6,371 km = 6,371,000 m).
- $ \theta $: Góc (hướng biên trái hoặc phải, tính bằng radian).
- $ \text{latitude} $: Vĩ độ hiện tại của camera.

In [17]:
def calculate_gps_coordinates(lat0, lon0, distance, angle, earth_radius=6_371_000):
    """
    Calculate GPS coordinates based on distance and angle.
    
    :param lat0: Initial latitude in degrees.
    :param lon0: Initial longitude in degrees.
    :param distance: Distance in meters.
    :param angle: Angle in degrees.
    :param earth_radius: Radius of the Earth in meters (default is 6,371,000 m).
    :return: New latitude and longitude in degrees.
    """
    angle_rad = radians(angle)
    lat0_rad = radians(lat0)
    lon0_rad = radians(lon0)
    
    delta_lat = (distance / earth_radius) * cos(angle_rad)
    new_lat = lat0_rad + delta_lat
    
    delta_lon = (distance / (earth_radius * cos(lat0_rad))) * sin(angle_rad)
    new_lon = lon0_rad + delta_lon
    
    return degrees(new_lat), degrees(new_lon)

In [18]:
# Function to calculate GPS coordinates based on distance and angle
def calculate_coordinates(lat, lon, dist, angle):
    """
    Calculate new GPS coordinates based on distance and angle.
    :param lat: Latitude of the starting point.
    :param lon: Longitude of the starting point.
    :param dist: Distance in meters.
    :param angle: Angle in degrees (bearing).
    :return: Tuple of new latitude and longitude.
    """
    point = distance(meters=dist).destination((lat, lon), bearing=angle)
    return point.latitude, point.longitude

Công thức $ D = \sqrt{2 \cdot R \cdot h} $ xuất phát từ hình học cầu, được dùng để tính khoảng cách từ một vị trí quan sát trên bề mặt Trái Đất đến đường chân trời.

1. **Mô hình hình học**:
   - Trái Đất được coi là một hình cầu với bán kính $ R $.
   - Một điểm quan sát ở độ cao $ h $ nằm ngoài bề mặt Trái Đất.
   - Đường chân trời được xác định là giao điểm của đường thẳng đi qua mắt quan sát và bề mặt Trái Đất.

2. **Phương trình hình học**:
   - Từ định lý Pythagoras, với tam giác vuông tạo bởi:
     - $ R $: Bán kính Trái Đất.
     - $ R + h $: Khoảng cách từ tâm Trái Đất đến điểm quan sát.
     - $ D $: Khoảng cách từ điểm quan sát đến đường chân trời.
   - Quan hệ: $(R + h)^2 = R^2 + D^2$.

3. **Đơn giản hóa**:
   - Vì $ h \ll R $ (độ cao nhỏ hơn rất nhiều so với bán kính Trái Đất), có thể xấp xỉ:
     $$
     R + h \approx R
     $$
   - Suy ra:
     $$
     D^2 \approx 2 \cdot R \cdot h
     $$
   - Do đó:
     $$
     D = \sqrt{2 \cdot R \cdot h}
     $$

In [26]:
# Get current GPS coordinates
#g = geocoder.ip('me')  # Get location based on IP address
#lat1, lon1 = g.latlng
lat1, lon1 = 43.57072605547293, 1.4682620342267738

# Increase distance for observation from a high floor
observation_distance = 20_000  # meters

# Calculate GPS coordinates for left and right boundaries
left_coordinates = calculate_gps_coordinates(lat1, lon1, observation_distance, left_angle)
right_coordinates = calculate_gps_coordinates(lat1, lon1, observation_distance, right_angle)

# Print results
print("Current GPS coordinates:", (lat1, lon1))
print("Left boundary GPS coordinates:", left_coordinates)
print("Right boundary GPS coordinates:", right_coordinates)

Current GPS coordinates: (43.57072605547293, 1.4682620342267738)
Left boundary GPS coordinates: (43.612623717370404, 1.2268395139826511)
Right boundary GPS coordinates: (43.73570966620624, 1.3693901517981564)


In [27]:
# Get current GPS coordinates
#g = geocoder.ip('me')  # Get location based on IP address
#lat1, lon1 = g.latlng
lat1, lon1 = 43.57072605547293, 1.4682620342267738

# Increase distance for observation from a high floor
observation_distance = 20_000  # meters

# Calculate GPS coordinates for left and right boundaries
left_coordinates = calculate_coordinates(lat1, lon1, observation_distance, left_angle)
right_coordinates = calculate_coordinates(lat1, lon1, observation_distance, right_angle)

# Print results
print("Current GPS coordinates:", (lat1, lon1))
print("Left boundary GPS coordinates:", left_coordinates)
print("Right boundary GPS coordinates:", right_coordinates)

Current GPS coordinates: (43.57072605547293, 1.4682620342267738)
Left boundary GPS coordinates: (43.612404145885904, 1.2273269095038501)
Right boundary GPS coordinates: (43.73579981703338, 1.3693876152922775)


In [28]:
def calculate_distance(lat1, lon1, lat2, lon2):
    R = 6371000  # Bán kính Trái Đất tính bằng mét

    # Chuyển đổi độ sang radian
    lat1_rad = math.radians(lat1)
    lon1_rad = math.radians(lon1)
    lat2_rad = math.radians(lat2)
    lon2_rad = math.radians(lon2)

    # Haversine formula
    dlat = lat2_rad - lat1_rad
    dlon = lon2_rad - lon1_rad
    a = math.sin(dlat/2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))

    distance = R * c
    return distance

In [35]:
print(calculate_distance(43.612623717370404, 1.2268395139826511, 43.612404145885904, 1.2273269095038501))
print(calculate_distance(43.73570966620624, 1.3693901517981564, 43.73579981703338, 1.3693876152922775))

46.2147371552092
10.026385853115254


In [51]:
from geopy.distance import geodesic

# Tọa độ của hai điểm (latitude, longitude)
point1 = left_coordinates
point2 = right_coordinates

# Tính khoảng cách
distance = geodesic(point1, point2).meters  # Đơn vị: mét
print(f"Khoảng cách: {distance} m")

Khoảng cách: 17866.476565779616 m


In [52]:
from haversine import haversine, Unit

# Tính khoảng cách
distance = haversine(point1, point2, unit=Unit.METERS)  # Đơn vị: mét
print(f"Khoảng cách: {distance} m")

Khoảng cách: 17855.017919284273 m


In [53]:
from pyproj import Geod

# Định nghĩa mô hình elipsoid WGS84
geod = Geod(ellps="WGS84")

# Tọa độ của hai điểm
lat1, lon1 = left_coordinates  # Hà Nội
lat2, lon2 = right_coordinates  # TP. HCM

# Tính khoảng cách
distance = geod.inv(lon1, lat1, lon2, lat2)[2]  # Khoảng cách ở index [2]
print(f"Khoảng cách: {distance} km")

Khoảng cách: 17866.476565779616 km


In [50]:
calculate_distance(lat1, lon1, lat2, lon2)

17854.993256917194

In [20]:
# Calculate boundary coordinates
boundary_points = []
for angle in range(int(left_angle), int(right_angle) + 1):
    boundary_points.append(calculate_gps_coordinates(lat1, lon1, observation_distance, angle))

# Add starting and ending points to close the polygon
boundary_points.insert(0, (lat1, lon1))
boundary_points.append((lat1, lon1))

# Create a folium map
m = folium.Map(location=[lat1, lon1], zoom_start=14)

# Add the contour area to the map
folium.Polygon(
    locations=boundary_points,
    color="blue",
    weight=2,
    fill=True,
    fill_color="blue",
    fill_opacity=0.4,
).add_to(m)

# Add a marker for the starting point
folium.Marker(
    [lat1, lon1],
    popup="Observation Point",
    icon=folium.Icon(color="red", icon="info-sign")
).add_to(m)

m

In [21]:
# Initial camera 1 coordinates and parameters
h = 10000  # Altitude in meters
FOV = 60  # Field of View in degrees
FOV_overlap = 20  # Overlap Field of View in degrees
center_orientation1 = 310  # Orientation of Camera 1 in degrees

# Calculate optimal distance
d = h * tan(radians(FOV / 2)) + h * tan(radians(FOV_overlap / 2))
d = 2 * h * tan(radians(FOV / 2))

# Calculate new camera 2 position
lat2, lon2 = calculate_gps_coordinates(lat1, lon1, d, center_orientation1)

# Calculate new camera 2 orientation
center_orientation2 = (center_orientation1 + 180) % 360

# Print results
print("Camera 2 Coordinates:", (lat2, lon2))
print("Camera 2 Orientation:", round(center_orientation2, 2), "degrees")
print(f"Optimal distance between cameras: {d / 1000:.2f} km")

Camera 2 Coordinates: (43.63747615112623, 1.3584662869373343)
Camera 2 Orientation: 130 degrees
Optimal distance between cameras: 11.55 km


In [22]:
# Calculate Camera 2 boundary
left_angle2 = center_orientation2 - (hfov / 2)
right_angle2 = center_orientation2 + (hfov / 2)
boundary_points2 = [calculate_gps_coordinates(lat2, lon2, observation_distance, angle) for angle in range(int(left_angle2), int(right_angle2) + 1)]
boundary_points2.insert(0, (lat2, lon2))
boundary_points2.append((lat2, lon2))

# Create folium map
m = folium.Map(location=[lat1, lon1], zoom_start=12)

# Add Camera 1 observation area
folium.Polygon(
    locations=boundary_points,
    color="blue",
    weight=2,
    fill=True,
    fill_color="blue",fill_opacity=0.4,
    popup="Camera 1 Area"
).add_to(m)

# Add Camera 2 observation area
folium.Polygon(
    locations=boundary_points2,
    color="green",
    weight=2,
    fill=True,
    fill_color="green",
    fill_opacity=0.4,
    popup="Camera 2 Area"
).add_to(m)

# Add markers for camera locations
folium.Marker([lat1, lon1], popup="Camera 1", icon=folium.Icon(color="red")).add_to(m)
folium.Marker([lat2, lon2], popup="Camera 2", icon=folium.Icon(color="blue")).add_to(m)

# Save and display map
m